sys/drv/bcm2835_pic.c

```diff
@@ -0,0 +1,204 @@
+                                   ih_filter_t *filter, ih_service_t *service,
+                                   void *arg, const char *name) {
+  bcm2835_pic_state_t *bcm2835_pic = pic->state;
+  unsigned irq = resource_start(r);
+  assert(irq < BCM2835_NIRQ);
+
+  if (bcm2835_pic->intr_event[irq] == NULL)
+    bcm2835_pic->intr_event[irq] =
+      intr_event_create(bcm2835_pic, irq, bcm2835_pic_disable_irq,
+                        bcm2835_pic_enable_irq, bcm2835_pic_intr_name(irq));
+
+  r->r_handler = intr_event_add_handler(bcm2835_pic->intr_event[irq], filter,
+                                        service, arg, name);
+}
+
+static void bcm2835_pic_teardown_intr(device_t *pic, device_t *dev,
+                                      resource_t *irq) {
+  intr_event_remove_handler(irq->r_handler);
+}
+
+static int bcm2835_pic_map_intr(device_t *pic, device_t *dev, phandle_t *intr,
+                                int icells) {
+  if (icells != 2)
+    return -1;
+
+  /* Interrupt tuples are of the form `<intrtype irq_offset>`. */
+  bcm2835_intrtype_t type = intr[0];
+  unsigned irq_off = intr[1];
+
+  if (type >= BCM2835_INTRTYPE_CNT)
+    return -1;
+  if (irq_off >= BCM2835_NIRQPERTYPE)
+    return -1;
+
+  unsigned irq = irq_off;
+  if (type == BCM2835_INTRTYPE_GPU1)
+    irq += BCM2835_NIRQPERTYPE;
+  else if (type == BCM2835_INTRTYPE_BASIC)
+    irq += BCM2835_NIRQPERTYPE * 2;
+  return irq;
+}
+
+static int bcm2835_pic_handle_intrtype(bcm2835_pic_state_t *bcm2835_pic,
+                                       bus_size_t reg, intr_event_t **events) {
+  uint32_t pending = in4(reg);
+  int rv = (pending != 0);
+
+  while (pending) {
+    unsigned irq = ffs(pending) - 1;
+    /* XXX: some pending bits are shared between BASIC and GPU0/1. */
+    if (events[irq])
+      intr_event_run_handlers(events[irq]);
+    pending &= ~(1 << irq);
+  }
+
+  return rv;
+}
+
+static intr_filter_t bcm2835_pic_intr_handler(void *arg) {
+  bcm2835_pic_state_t *bcm2835_pic = arg;
+
+  int gpu0_intr = bcm2835_pic_handle_intrtype(
+    bcm2835_pic, BCM2835_INTC_IRQ1PENDING, bcm2835_pic->intr_event);
+  int gpu1_intr =
+    bcm2835_pic_handle_intrtype(bcm2835_pic, BCM2835_INTC_IRQ2PENDING,
+                                &bcm2835_pic->intr_event[BCM2835_INT_GPU1BASE]);
+  int basic_intr = bcm2835_pic_handle_intrtype(
+    bcm2835_pic, BCM2835_INTC_IRQBPENDING,
+    &bcm2835_pic->intr_event[BCM2835_INT_BASICBASE]);
+
+  return (gpu0_intr || gpu1_intr || basic_intr) ? IF_FILTERED : IF_STRAY;
+}
+
+static int bcm2835_pic_probe(device_t *pic) {
+  return FDT_is_compatible(pic->node, "brcm,bcm2836-armctrl-ic");
+}
+
+static int bcm2835_pic_attach(device_t *pic) {
+  bcm2835_pic_state_t *bcm2835_pic = pic->state;
+
+  rman_init(&bcm2835_pic->rm, "BCM2835 PIC interrupt sources");
+  rman_manage_region(&bcm2835_pic->rm, 0, BCM2835_NIRQ);
+
+  bcm2835_pic->mem = device_take_memory(pic, 0, RF_ACTIVE);
+  assert(bcm2835_pic->mem);
```

Comment:
For future - we don't handle errors of this function and always require that return value is valid. Maybe we should explicitly said that this function never fail and move assert to body.

Reply:
I'm not sure. @cahirwpz what do you think?